consensus/istanbul/core/roundstate_db.go

```diff
@@ -19,35 +19,75 @@ package core
 import (
 	"bytes"
 	"encoding/binary"
+	"math/big"
 	"os"
+	"time"
 
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/task"
 	"github.com/ethereum/go-ethereum/consensus/istanbul"
 	"github.com/ethereum/go-ethereum/log"
 	"github.com/ethereum/go-ethereum/rlp"
 	"github.com/syndtr/goleveldb/leveldb"
 	lvlerrors "github.com/syndtr/goleveldb/leveldb/errors"
 	"github.com/syndtr/goleveldb/leveldb/opt"
 	"github.com/syndtr/goleveldb/leveldb/storage"
+	"github.com/syndtr/goleveldb/leveldb/util"
 )
 
 const (
-	dbVersion    = 1
+	dbVersion    = 2
 	dbVersionKey = "version"  // Version of the database to flush if changes
 	lastViewKey  = "lastView" // Last View that we know of
+	rsKey        = "rs"       // Database Key Pefix for RoundState
 )
 
 type RoundStateDB interface {
 	GetLastView() (*istanbul.View, error)
+	// GetOldestValidView returns the oldest valid view that can be stored on the db
+	// it might or might not be present on the db
+	GetOldestValidView() (*istanbul.View, error)
 	GetRoundStateFor(view *istanbul.View) (RoundState, error)
 	UpdateLastRoundState(rs RoundState) error
 	Close() error
 }
 
+// RoundStateDBOptions are the options for a RoundStateDB instance
+type RoundStateDBOptions struct {
+	withGarbageCollector   bool
+	garbageCollectorPeriod time.Duration
+	sequencesToSave        uint64
+}
+
 type roundStateDBImpl struct {
-	db *leveldb.DB
+	db                   *leveldb.DB
+	stopGarbageCollector task.StopFn
+	opts                 RoundStateDBOptions
+	logger               log.Logger
+}
+
+var defaultRoundStateDBOptions = RoundStateDBOptions{
+	withGarbageCollector:   true,
+	sequencesToSave:        100,
```

Comment:
For the future, it will need to save sequences where a validator committed multiple blocks for far longer, namely the duration of the statute of limitations. I think it's fine to save this for when that type of slashing gets implemented.